zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -49,12 +49,20 @@ pub struct Chain {
     /// The Sprout note commitment tree of the tip of this `Chain`,
     /// including all finalized notes, and the non-finalized notes in this chain.
     pub(super) sprout_note_commitment_tree: sprout::tree::NoteCommitmentTree,
+    /// The Sprout note commitment tree for each anchor.
+    /// This is required for interstitial states.
+    pub(crate) sprout_trees_by_anchor:
+        HashMap<sprout::tree::Root, sprout::tree::NoteCommitmentTree>,
     /// The Sapling note commitment tree of the tip of this `Chain`,
     /// including all finalized notes, and the non-finalized notes in this chain.
     pub(super) sapling_note_commitment_tree: sapling::tree::NoteCommitmentTree,
+    /// The Sapling note commitment tree for each height.
+    pub(crate) sapling_trees_by_height: BTreeMap<block::Height, sapling::tree::NoteCommitmentTree>,
     /// The Orchard note commitment tree of the tip of this `Chain`,
     /// including all finalized notes, and the non-finalized notes in this chain.
     pub(super) orchard_note_commitment_tree: orchard::tree::NoteCommitmentTree,
+    /// The Orchard note commitment tree for each height.
+    pub(crate) orchard_trees_by_height: BTreeMap<block::Height, orchard::tree::NoteCommitmentTree>,
```

Comment:
Performance/Memory: This adds about 10 kB of data to each 100-block `Chain`. We clone this chain data for every block.

That might be ok, but we need to run a manual full sync test to make sure it's not too slow. I can do that after I finish my full sync tests.

If it is too slow, I added some more optimisations to #3757.

Reply:
This is running at https://github.com/ZcashFoundation/zebra/actions/runs/1961144403

Getting to around 99.99% by 6 hours is good enough, but we don't want it to be much slower.

Reply:
Looks good:

```
Mar 10 04:45:23.455  INFO ***net="Main"***: zebrad::commands::start: Starting zebrad
...
Mar 10 09:58:28.656  INFO ***net="Main"***: zebrad::commands::start: estimated progress to chain tip sync_percent=99.083 % current_height=Height(1578224) remaining_sync_blocks=14610 time_since_last_state_block=PT0S
...
Mar 10 10:43:28.709  INFO ***net="Main"***: zebrad::commands::start: estimated progress to chain tip sync_percent=99.990 % current_height=Height(1592649) remaining_sync_blocks=162 time_since_last_state_block=-PT300.005696497S
```

Reply:
It may be possible to use a height -> anchor map instead so the tree can be looked up by height -> anchor -> tree (we already seem to do that for Sprout). But it may not be worth the hassle since it did not seem to impact performance.

Reply:
I'd like to use `Arc<Tree>` and `Arc<Root>` to make cloning cheaper, because might be cloning `Chain`s a bit more in #3835.

But I'll add that to #3757.